strawberryfields/engine.py

```diff
@@ -168,11 +174,108 @@ def _run_program_locally(self, prog, **kwargs):
         Returns:
             list[Command]: commands that were applied to the backend
         """
+
+    def _run(self, program, *, compile=True, **kwargs):
+        """Execute the given programs by sending them to the backend.
+
+        For each :class:`Program` instance given as input, the following happens:
+
+        * The Program instance is compiled and optimized.
+        * The compiled program is executed on the backend, updating the backend state.
+        * The latest measurement results of each subsystem are stored
+          in the :class:`.RegRef` instances of the corresponding Program, as well as in self.samples.
+        * The compiled program is appended to self.run_progs.
+
+        Finally, the result of the computation is returned.
+
+        Args:
+            program (Program, Sequence[Program]): quantum programs to run
+            compile (bool): If True, compile the Program instances before sending them to the backend.
+
+        The ``kwargs`` keyword arguments are passed to :meth:`_run_program`.
+
+        Returns:
+            Result: results of the computation
+        """
+        if not isinstance(program, Sequence):
+            program = [program]
+
+        try:
+            prev = self.run_progs[-1] if self.run_progs else None  # previous program segment
+            for p in program:
+                if prev is None:
+                    # initialize the backend
+                    self._init_backend(p.init_num_subsystems)
+                else:
+                    # there was a previous program segment
+                    if not p.can_follow(prev):
+                        raise RuntimeError("Register mismatch: program {}, '{}'.".format(len(self.run_progs), p.name))
+
+                    # Copy the latest measured values in the RegRefs of p.
+                    # We cannot copy from prev directly because it could be used in more than one engine.
+                    for k, v in enumerate(self.samples):
+                        p.reg_refs[k].val = v
+
+                # if the program hasn't been compiled for this backend, do it now
+                if compile and p.backend != self.backend_name:
```

Comment:
Maybe the engine should compile it if it hasn't been compiled yet for the targeted backend? I.e. remove "compile and" here.

Reply:
Yep, that would work

Reply:
I just remembered, the reason it's required is that some tests fail if the program is compiled (because compilation can change the order of gates that commute). Maybe we need more nuanced compilation options like "preserve gate order".

Reply:
Ah, that's annoying. I guess it doesn't matter too much for local engines, but for remote engines we can write the tests to get around this problem.